src/Ouroboros-Mini_Protocols-Chain_Sync.thy

```diff
@@ -124,30 +135,43 @@ definition index :: "('i \<Rightarrow> 'q) \<Rightarrow> 'q \<Rightarrow> 'i lis
 definition first_intersection_point :: "('i \<Rightarrow> 'q) \<Rightarrow> 'q list \<Rightarrow> 'i list \<rightharpoonup> 'q" where
   [simp]: "first_intersection_point \<psi> qs C  = find (\<lambda>q. q \<in> \<psi> ` set C) qs"
 
+definition last_intersection_point :: "('i \<Rightarrow> 'q) \<Rightarrow> 'i list \<Rightarrow> 'i list \<Rightarrow> 'q" where
+  [simp]: "last_intersection_point \<psi> C C' = (ARG_MAX (\<lambda>q. index \<psi> q C) q. q \<in> \<psi> ` (set C \<inter> set C'))"
+
 corec server_program where
-  "server_program \<psi> C k b =
+  "server_program \<psi> C k b a =
     \<down> M. (partial_case M of
       Done \<Rightarrow>
         \<bottom> |
       Cont (FindIntersect qs) \<Rightarrow>
         (case first_intersection_point \<psi> qs C of
           None \<Rightarrow>
             \<up> Cont IntersectNotFound;
-            server_program \<psi> C k b |
+            server_program \<psi> C k b a |
           Some q \<Rightarrow>
             \<up> Cont (IntersectFound q);
-            server_program \<psi> C (index \<psi> q C) True
+            server_program \<psi> C (index \<psi> q C) True a
         ) |
       Cont RequestNext \<Rightarrow>
         if b then
           \<up> Cont (RollBackward (\<psi> (C ! k)));
-          server_program \<psi> C k False
+          server_program \<psi> C k False a
         else if Suc k < length C then
           \<up> Cont (RollForward (C ! Suc k));
-          server_program \<psi> C (Suc k) b
-        else \<comment> \<open>only for this initial implementation\<close>
+          server_program \<psi> C (Suc k) b a
+        else
           \<up> Cont AwaitReply;
-          server_program \<psi> C k b
+          a \<rightarrow> C'. (
+            \<comment> \<open>assumed that at least \<^term>\<open>C\<close>~and~\<^term>\<open>C'\<close> coincide in the genesis block\<close>
+            let q = last_intersection_point \<psi> C C' in
+            if q = \<psi> (last C) then
+              \<comment> \<open>assumed that \<^term>\<open>C'\<close> is longer than \<^term>\<open>C\<close>\<close>
+              \<up> Cont (RollForward (C' ! Suc k));
+              server_program \<psi> C' (Suc k) b a
+            else
+              \<up> Cont (RollBackward q);
+              server_program \<psi> C' (index \<psi> q C') b a
+          )
     )"
```

Comment:
There is an issue with the algorithm that the current version of `server_program` implements: it only checks for new chains when the client has caught up. A correct algorithm would check for new chains also when being in the `CanAwait` state at least.

However, we can’t have a mini-protocol program check whether some process is prepared to send a value to a particular synchronous channel. Therefore, there is no way for the server program to check whether a new chain is provided if providing new chains is directly implemented by message passing. What would be good to have here is really some sort of read access to a shared state, which contains the current chain. This way, the server program could always get a current chain (no blocking possible), which it could then compare to the chain used so far to detect chain changes.

Interestingly, it seems to be possible to “abuse” the synchronous communication mechanism for realizing read access to mutable state, although this relies on the particularities of the implementation of synchronous channels. Receiving from a synchronous channel is implemented as sending a one-time, private channel and then receiving from this channel. Sending, on the other hand, is implemented as receiving such a one-time, private channel and then sending to it. To realize read access to mutable state, it is enough to avoid ordinary sending and instead run a process that repeatedly receives a one-time, private channel, then reads the mutable state, and finally sends the acquired state to the one-time, private channel. Nothing has to change on the receiver’s end: ordinary synchronous receiving can be used to get the current state.

Do you think we should “abuse” synchronous communication in this way and implement regular checks for new chain data this way?

Reply:
Yes, I think we should do that. Would we need to add something new to our framework?

Reply:
Interestingly, in _Communication and Mobile Systems: The π-Calculus_ (Milner, 1999) a technique is developed to represent mutable state on top of the (synchronous) π-calculus. It may be an overkill for our purposes but nevertheless I think it's worth having a look.

Reply:
Sounds all very good.

I don’t think we need to change the framework at this point. We can just use a synchronous channel directly.

Reply:
When using (emulated) mutable state, we should acquire the initial server chain by reading the state. As a result, we should drop the `initial_server_chain` parameter.

Reply:
Fixed in 4b6488c47e1500f3f69dfab23af37680da18676a, please check.

Reply:
https://github.com/input-output-hk/ouroboros-high-assurance/blob/4b6488c47e1500f3f69dfab23af37680da18676a/src/Ouroboros-Mini_Protocols-Chain_Sync.thy#L135-L136

I think that the `C`- and the `k`-argument should be swapped, so that their order is in line with the argument order of `server_chain'`.

Reply:
https://github.com/input-output-hk/ouroboros-high-assurance/blob/4b6488c47e1500f3f69dfab23af37680da18676a/src/Ouroboros-Mini_Protocols-Chain_Sync.thy#L137-L143

The parentheses at the outside are superfluous and thus should be dropped.

Reply:
https://github.com/input-output-hk/ouroboros-high-assurance/blob/4b6488c47e1500f3f69dfab23af37680da18676a/src/Ouroboros-Mini_Protocols-Chain_Sync.thy#L138-L142

I think that the logic here is wrong: If `C` is not a prefix of `C'` but the read pointer points into the longest common prefix of `C` and `C'`, then the above code results in a `RollBackward` to the end of the longest common prefix, which usually means making a leap forward. I guess that the distinction should not be between `C` being a prefix of `C'` or not but between the read pointer pointing into the longest common prefix or not (`C` being a prefix of `C'` is a special case of the read pointer pointing into the longest common prefix).

Reply:
https://github.com/input-output-hk/ouroboros-high-assurance/blob/4b6488c47e1500f3f69dfab23af37680da18676a/src/Ouroboros-Mini_Protocols-Chain_Sync.thy#L135-L136

I think that this function should get a different name. The current name isn’t really descriptive, as it doesn’t say how the delivered function results relate to a “chain switch”. Furthermore, the function is not only about switching chains, as it also handles the case where the chain is extended.

Reply:
https://github.com/input-output-hk/ouroboros-high-assurance/blob/4b6488c47e1500f3f69dfab23af37680da18676a/src/Ouroboros-Mini_Protocols-Chain_Sync.thy#L135-L143

It might make sense to include the case `C = C'` into this function. First, this case is distinguished in both phases. Second, without this case being handled here, the `chain_switch` function should be unspecified for this case (via a clause `if C ≠ C'`), but this would make in particular the conformance proof more difficult, I guess. Since there is no common reaction to the situation that the chain hasn’t changed among the two phases, the `cain_sync` function could just yield a special value in this case.

Reply:
https://github.com/input-output-hk/ouroboros-high-assurance/blob/4b6488c47e1500f3f69dfab23af37680da18676a/src/Ouroboros-Mini_Protocols-Chain_Sync.thy#L178-L196

*The Shelley Networking Protocol* says the following:

> Whenever the server replies with `MsgIntersectFound` the client can expect the next update (i.e. a replay to `MsgRequestNext`) to be `MsgRollBackward` to the specified *point_intersect* (which makes handling state updates on the client side easier).

However, the above code exhibits this behavior only when the server chain hasn’t changed.

Reply:
https://github.com/input-output-hk/ouroboros-high-assurance/blob/4b6488c47e1500f3f69dfab23af37680da18676a/src/Ouroboros-Mini_Protocols-Chain_Sync.thy#L199-L200

We do not need to have the arguments `u`, `k`, and `φ` here, as they can only have the values `0`, `False`, and `ClientLagging` anyhow. We should just drop these arguments and pass said initial values to `server_program'` as part of the definition of `server_program`. Actually, I think that we should do something similar with the client program: have a wrapper function that passes fixed initial arguments.

Reply:
https://github.com/input-output-hk/ouroboros-high-assurance/blob/4b6488c47e1500f3f69dfab23af37680da18676a/src/Ouroboros-Mini_Protocols-Chain_Sync.thy#L153

I think that the identifier `server_program'` is problematic. First, this is not the whole server program but only its main part. Second, we alreay use identifiers with apostrophes for a completely different purpose: for concrete values of state machine fields. What about `server_main_loop`?

Reply:
https://github.com/input-output-hk/ouroboros-high-assurance/blob/4b6488c47e1500f3f69dfab23af37680da18676a/src/Ouroboros-Mini_Protocols-Chain_Sync.thy#L145-L147

I think that the names of the phases should be changed. First, they are not consistent with the names for client phases, which describe activities. Second, the name `ClientCatchUp` is misleading, because it doesn’t refer to the client catching up, but on the contrary to the client already having *caught* up. Unfortunately, I don’t know any good alternatives at the moment.

Reply:
https://github.com/input-output-hk/ouroboros-high-assurance/blob/4b6488c47e1500f3f69dfab23af37680da18676a/src/Ouroboros-Mini_Protocols-Chain_Sync.thy#L155-L196

The two cases should be handled in opposite order, to conform with the order the respective phases are declared.

Reply:
https://github.com/input-output-hk/ouroboros-high-assurance/blob/4b6488c47e1500f3f69dfab23af37680da18676a/src/Ouroboros-Mini_Protocols-Chain_Sync.thy#L158
https://github.com/input-output-hk/ouroboros-high-assurance/blob/4b6488c47e1500f3f69dfab23af37680da18676a/src/Ouroboros-Mini_Protocols-Chain_Sync.thy#L160
https://github.com/input-output-hk/ouroboros-high-assurance/blob/4b6488c47e1500f3f69dfab23af37680da18676a/src/Ouroboros-Mini_Protocols-Chain_Sync.thy#L181
https://github.com/input-output-hk/ouroboros-high-assurance/blob/4b6488c47e1500f3f69dfab23af37680da18676a/src/Ouroboros-Mini_Protocols-Chain_Sync.thy#L188
https://github.com/input-output-hk/ouroboros-high-assurance/blob/4b6488c47e1500f3f69dfab23af37680da18676a/src/Ouroboros-Mini_Protocols-Chain_Sync.thy#L191

I would drop these comments, since we also don’t comment elsewhere in our programs and rely on these programs being self-explanatory (which they should be, because otherwise they would be bad as high-level specifications).

Reply:
https://github.com/input-output-hk/ouroboros-high-assurance/blob/4b6488c47e1500f3f69dfab23af37680da18676a/src/Ouroboros-Mini_Protocols-Chain_Sync.thy#L149-L151

I think that the name `server_state_in_phase` is misleading, since it suggest to talk about a “server state”, which doesn’t exist as states refer to the whole system on server and client, and a “phase”, which would be ambiguous, as there are server phases and client phases. What about `state_in_server_phase`?

Reply:
https://github.com/input-output-hk/ouroboros-high-assurance/blob/4b6488c47e1500f3f69dfab23af37680da18676a/src/Ouroboros-Mini_Protocols-Chain_Sync.thy#L200

The comment in this line should be removed. The chain not being empty is covered by the constraint that the channel with chain updates only yields non-empty chains, and documenting this is part of what this review calls for.

Reply:
https://github.com/input-output-hk/ouroboros-high-assurance/blob/4b6488c47e1500f3f69dfab23af37680da18676a/src/Ouroboros-Mini_Protocols-Chain_Sync.thy#L230

There should be a space before `⦇`.

Reply:
https://github.com/input-output-hk/ouroboros-high-assurance/blob/4b6488c47e1500f3f69dfab23af37680da18676a/src/Ouroboros-Mini_Protocols-Chain_Sync.thy#L241-L245

This looks like it should be improved, but since the handling of wrappers around main loops and these things will probably be changed anyhow (see [above](https://github.com/input-output-hk/ouroboros-high-assurance/pull/93#discussion_r1398526533)), this part might be subject to change. I’ll properly review any future version of this later.

Reply:
Let me follow up on what I wrote [above](https://github.com/input-output-hk/ouroboros-high-assurance/pull/93#discussion_r1398250892), where I explained that the current algorithm for determining the next step (waiting, rolling forward, or rolling backward) is not correct.

A correct algorithm for deciding what to do next is the following one: First, we form the longest common prefix of the old and the new chain. Then we continue according to how the read pointer is related to the length of this prefix and the length of the new chain:
* If the read pointer is smaller than the prefix’s length and smaller than the predecessor of the new chain’s length, then we roll forward.
* If the read pointer is smaller than the prefix’s length and just one less than the new chain’s length, then we wait.
* If the read pointer is equal to or greater than the prefix’s length, then we roll backward.

Note that this algorithm does not check whether the old and the new chain are equal or not. In fact, this seems to be irrelevant. What is actually relevant is whether the read pointer points into the longest common prefix of the old and the new chain.

However, this is equivalent to whether the prefix of the old chain that ends with the item to which the read pointer points is a prefix of the new chain. This suggests a change to the internal state of the main loop (which is represented by arguments): instead of maintaining the old chain, of which we only need said prefix, and additionally the read pointer, we may just maintain said prefix. No need for a read pointer then, since it would only be the predecessor of the length of the prefix. Note that said prefix is not just an artificial construct but has a straightforward interpretation: it is the part of the chain that the client possesses (or will possess after having processed any messages currently in transit).

With this change of internal state, the server program would work as follows: Initially, it reads the current server chain and invokes the main loop with the list that only contains the genesis item. In each iteration of the main loop, the next step is chosen as follows:
* If the maintained prefix is a proper prefix of the new chain, then we roll forward.
* If the maintained prefix is equal to the new chain, then we wait.
* If the maintained prefix is neither of the two, then we pick the longest common prefix of the maintained prefix and the new chain and roll backward.

I think it is worthwhile to implement this final algorithm, not least because it is simpler than the one sketched above.

Reply:
> the `C`- and the `k`-argument should be swapped

Fixed in 3b0db94d122b2a43e2ccf20fb1191b0fd6b66d57, please check.

Reply:
> https://github.com/input-output-hk/ouroboros-high-assurance/blob/4b6488c47e1500f3f69dfab23af37680da18676a/src/Ouroboros-Mini_Protocols-Chain_Sync.thy#L137-L143
> 
> I think that this whole expression should be indented relative to the beginning of the equation above it.

Fixed in 6e4963d5f8ce5e202db5a3810814209aa68c8df4, please check.

Reply:
> https://github.com/input-output-hk/ouroboros-high-assurance/blob/4b6488c47e1500f3f69dfab23af37680da18676a/src/Ouroboros-Mini_Protocols-Chain_Sync.thy#L199-L200
> 
> We do not need to have the arguments `u`, `k`, and `φ` here, as they can only have the values `0`, `False`, and `ClientLagging` anyhow. We should just drop these arguments and pass said initial values to `server_program'` as part of the definition of `server_program`. Actually, I think that perhaps we should do something similar with the client program: have a wrapper function that passes fixed initial arguments. An alternative approach would be to have wrapper functions for neither the server nor the client and instead put the passing of fixed initial arguments as well as the initial reading of the server chain into the definition of `program`.

I think the alternative approach you propose above is the most appropriate one. Implemented in ca417ee3565d9f506c28cf55ec21451c7d2a64ce, please check.

Reply:
> https://github.com/input-output-hk/ouroboros-high-assurance/blob/4b6488c47e1500f3f69dfab23af37680da18676a/src/Ouroboros-Mini_Protocols-Chain_Sync.thy#L135-L136
> 
> I think that this function should get a different name. The current name isn’t really descriptive, as it doesn’t say how the delivered function results relate to a “chain switch”. Furthermore, the function is not only about switching chains, as it also handles the case where the chain is extended.

Fixed in 517eb1b6de7919c44f9287d37e909955d5abac29, please check.

Reply:
> https://github.com/input-output-hk/ouroboros-high-assurance/blob/4b6488c47e1500f3f69dfab23af37680da18676a/src/Ouroboros-Mini_Protocols-Chain_Sync.thy#L158
> 
> https://github.com/input-output-hk/ouroboros-high-assurance/blob/4b6488c47e1500f3f69dfab23af37680da18676a/src/Ouroboros-Mini_Protocols-Chain_Sync.thy#L160
> 
> https://github.com/input-output-hk/ouroboros-high-assurance/blob/4b6488c47e1500f3f69dfab23af37680da18676a/src/Ouroboros-Mini_Protocols-Chain_Sync.thy#L181
> 
> https://github.com/input-output-hk/ouroboros-high-assurance/blob/4b6488c47e1500f3f69dfab23af37680da18676a/src/Ouroboros-Mini_Protocols-Chain_Sync.thy#L188
> 
> https://github.com/input-output-hk/ouroboros-high-assurance/blob/4b6488c47e1500f3f69dfab23af37680da18676a/src/Ouroboros-Mini_Protocols-Chain_Sync.thy#L191
> 
> I would drop these comments, since we also don’t comment elsewhere in our programs and rely on these programs being self-explanatory (which they should be, because otherwise they would be bad as high-level specifications).

Fixed in 37b475da428f5d3af619751edadc5038f76d8b3a, please check.

Reply:
> https://github.com/input-output-hk/ouroboros-high-assurance/blob/4b6488c47e1500f3f69dfab23af37680da18676a/src/Ouroboros-Mini_Protocols-Chain_Sync.thy#L200
> 
> The comment in this line should be removed. The chain not being empty is covered by the constraint that the channel with chain updates only yields non-empty chains, and documenting this is part of what this review calls for.

Fixed in b4a8654c1629b6569f9e5c890eb595a45cd4c493, please check.

Reply:
> https://github.com/input-output-hk/ouroboros-high-assurance/blob/4b6488c47e1500f3f69dfab23af37680da18676a/src/Ouroboros-Mini_Protocols-Chain_Sync.thy#L230
> 
> There should be a space before `⦇`.

Fixed in 9012e34ca15b6746ab6bb09bb9c5e4129db11e5e, please check.

Reply:
> With this change of internal state, the server program would work as follows: Initially, it reads the current server chain and invokes the main loop with the list that only contains the genesis item. In each iteration of the main loop, the next step is chosen as follows:
> 
>   * If the maintained prefix is a proper prefix of the new chain, then we roll forward.
>   * If the maintained prefix is equal to the new chain, then we wait.
>   * If the maintained prefix is neither of the two, then we pick the longest common prefix of the maintained prefix and the new chain and roll backward.
> 
> 
> I think it is worthwhile to implement this final algorithm, not least because it is simpler than the one sketched above.

This is done now as part of 1c6a24f7538f33fddc60816d3b29ae0044305caa.

Reply:
> https://github.com/input-output-hk/ouroboros-high-assurance/blob/4b6488c47e1500f3f69dfab23af37680da18676a/src/Ouroboros-Mini_Protocols-Chain_Sync.thy#L178-L196
> 
> _The Shelley Networking Protocol_ says the following:
> 
> > Whenever the server replies with `MsgIntersectFound` the client can expect the next update (i.e. a replay to `MsgRequestNext`) to be `MsgRollBackward` to the specified _point_intersect_ (which makes handling state updates on the client side easier).
> 
> However, the above code exhibits this behavior only when the server chain hasn’t changed.

This has been fixed as part of 1c6a24f7538f33fddc60816d3b29ae0044305caa.

Reply:
> https://github.com/input-output-hk/ouroboros-high-assurance/blob/4b6488c47e1500f3f69dfab23af37680da18676a/src/Ouroboros-Mini_Protocols-Chain_Sync.thy#L241-L245
> 
> This looks like it should be improved, but since the handling of wrappers around main loops and these things will probably be changed anyhow (see [above](https://github.com/input-output-hk/ouroboros-high-assurance/pull/93#discussion_r1398526533)), this part might be subject to change. I’ll properly review any future version of this later.

This code section has been improved as part of 1c6a24f7538f33fddc60816d3b29ae0044305caa.

Reply:
> https://github.com/input-output-hk/ouroboros-high-assurance/blob/4b6488c47e1500f3f69dfab23af37680da18676a/src/Ouroboros-Mini_Protocols-Chain_Sync.thy#L135-L143
> 
> It might make sense to include the case `C = C'` into this function.

This function has been overhauled as part of 1c6a24f7538f33fddc60816d3b29ae0044305caa, and the new version handles all cases.

Reply:
> https://github.com/input-output-hk/ouroboros-high-assurance/blob/4b6488c47e1500f3f69dfab23af37680da18676a/src/Ouroboros-Mini_Protocols-Chain_Sync.thy#L145-L147
> 
> I think that the names of the phases should be changed. First, they are not consistent with the names for client phases, which describe activities. Second, the name `ClientCatchUp` is misleading, because it doesn’t refer to the client catching up, but on the contrary to the client already having _caught_ up. Unfortunately, I don’t know any good alternatives at the moment.

This has been fixed in bdfe40e58065b2ea608998c0765412a724ab64e4.

Reply:
Now that `server_chain_updates` has been renamed to `server_chains` it might make sense to change the name of the argument `u` (perhaps to `c`?)

Reply:
I was thinking about that. However, `C` is used for a single chain; so for a channel of chains we should rather use something like `Cs` instead of `C`. Then again, `Cs` looks more like an identifier for a *list* of chains. Should we use it nevertheless?

Reply:
I don't like `Cs` for the same reason you stated above. Since I cannot think of a better alternative to `u` now, I think we can safely leave is as it currently is.

Reply:
Hmm, but I also don’t like the `u` anymore. :slightly_smiling_face:

Reply:
As we discussed outside GitHub, `𝒞` seems like the way to go. I made a corresponding change in 7be24440ef89a60080a883a6c60924d72a010806.